File: wrkspc/kana-core/src/parser/fsm.rs
use crate::transliterate::CharFeature;

use super::{glyphs::punctuation::MARKS as PUNCTUATION_MARKS, input::Input, utf8};

#[derive(Clone, Copy, Debug)]
pub enum State<'a> {
	Init,
	RawToggle,
	KanaToggle,
	SizeRouter,
	Long,
	Medium,
	Short,
	Tiny,
	Sokuon,
	VirtualSokuon,
	SmallVowel,
	Chouonpu,
	RawText(char),
	Punctuation(usize, &'a Self),
	Fallback,
}

impl<'a> State<'a> {
	pub fn next(self, mut input: Input<'a>) -> Option<(&'a str, Input<'a>, State)> {
		let size = self.get_size();
		let romaji = input.romaji;

		match self {
			Self::Init => match romaji.is_empty() {
				true => None,
				false => Some(("", input, Self::RawToggle)),
			},
			state => Some(match state {
				Self::RawToggle => match input.special_chars.get(&CharFeature::RawTextToggle) {
					Some(toggle) if romaji.starts_with(*toggle) => (
						"",
						{
							input.romaji = utf8::slice_from(romaji, size);
							input
						},
						Self::RawText(*toggle),
					),
					_ => ("", input, Self::KanaToggle),
				},
				Self::RawText(toggle) => {
					input.romaji = utf8::slice_from(romaji, size);

					match romaji.starts_with(toggle) {
						true => ("", input, Self::RawToggle),
						false => (
							utf8::slice_to(romaji, size),
							input,
							if romaji.is_empty() {
								Self::Init
							} else {
								Self::RawText(toggle)
							},
						),
					}
				}
				Self::KanaToggle => match input.special_chars.get(&CharFeature::KanaToggle) {
					Some(toggle) if input.romaji.starts_with(*toggle) => (
						"",
						{
							input.kanas.toggle();
							input.romaji = utf8::slice_from(romaji, size);
							input
						},
						Self::Init,
					),
					_ => ("", input, Self::SizeRouter),
				},
				Self::SizeRouter => (
					"",
					input,
					match utf8::count_chars(romaji) {
						n if n >= 4 => Self::Long,
						3 => Self::Medium,
						2 => Self::Sokuon,
						1 => Self::Tiny,
						_ => Self::Init,
					},
				),
				Self::Long => {
					let selection = utf8::slice_to(romaji, size);

					match input.kanas.get_current().get(selection) {
						Some(output) => (
							output,
							{
								input.romaji = utf8::slice_from(romaji, size - 1);
								input
							},
							Self::Chouonpu,
						),
						None => ("", input, Self::Punctuation(size, &Self::Medium)),
					}
				}
				Self::Medium => {
					let selection = utf8::slice_to(romaji, size);

					match input.kanas.get_current().get(selection) {
						Some(output) => (
							output,
							{
								input.romaji = utf8::slice_from(romaji, size - 1);
								input
							},
							Self::Chouonpu,
						),
						None => ("", input, Self::Punctuation(size, &Self::Sokuon)),
					}
				}
				Self::Sokuon => {
					let selection = utf8::slice_to(romaji, size);

					match input.kanas.get_current().sokuon(selection) {
						Some(output) => (
							output,
							{
								input.romaji = utf8::slice_from(romaji, size - 1);
								input
							},
							Self::Init,
						),
						None => ("", input, Self::SmallVowel),
					}
				}
				Self::SmallVowel => {
					let selection = utf8::slice_to(romaji, size);

					match input.special_chars.get(&CharFeature::SmallVowel) {
						Some(trigger) if selection.starts_with(*trigger) => {
							match input
								.kanas
								.get_current()
								.small_vowel(&utf8::slice_from(selection, 1))
							{
								Some(vowel) => (
									vowel,
									{
										input.romaji = utf8::slice_from(romaji, size - 1);
										input
									},
									Self::Chouonpu,
								),
								_ => ("", input, Self::Short),
							}
						}
						_ => ("", input, Self::Short),
					}
				}
				Self::Short => {
					let selection = utf8::slice_to(romaji, size);

					match input.kanas.get_current().get(selection) {
						Some(output) => (
							output,
							{
								input.romaji = utf8::slice_from(romaji, size - 1);
								input
							},
							Self::Chouonpu,
						),
						None => ("", input, Self::Punctuation(size, &Self::Tiny)),
					}
				}
				Self::Tiny => {
					let selection = utf8::slice_to(romaji, size);

					match input.kanas.get_current().get(selection) {
						Some(output) => (output, input, Self::Chouonpu),
						None => ("", input, Self::Punctuation(size, &Self::VirtualSokuon)),
					}
				}
				Self::Chouonpu => match input.special_chars.get(&CharFeature::Reset) {
					Some(reset) if utf8::slice_from(romaji, 1).starts_with(*reset) => (
						"",
						{
							input.romaji = utf8::slice_from(romaji, size);
							input
						},
						Self::Init,
					),
					_ => {
						let selection = utf8::slice_to(romaji, size);
						input.romaji = utf8::slice_from(romaji, size - 1);

						match input.kanas.get_current().chouonpu(selection) {
							Some(output) => (output, input, Self::Chouonpu),
							None => ("", input, Self::Init),
						}
					}
				},
				Self::Punctuation(size, next) => {
					let selection = utf8::slice_to(romaji, size);

					match input.punctuation {
						Some(ref mut punctuation) => match selection {
							"'" => {
								let output = punctuation.single_quotes.get_current();

								(
									output,
									{
										punctuation.single_quotes.toggle();
										input.romaji = utf8::slice_from(romaji, 1);
										input
									},
									Self::Init,
								)
							}
							r#"""# => {
								let output = punctuation.double_quotes.get_current();

								(
									output,
									{
										punctuation.double_quotes.toggle();
										input.romaji = utf8::slice_from(romaji, 1);
										input
									},
									Self::Init,
								)
							}
							_ => match PUNCTUATION_MARKS.get(selection) {
								Some(output) => (
									output,
									{
										input.romaji = utf8::slice_from(romaji, size);
										input
									},
									Self::Init,
								),
								None => ("", input, *next),
							},
						},
						None => ("", input, *next),
					}
				}
				Self::VirtualSokuon => match input.special_chars.get(&CharFeature::VirtualSokuon) {
					Some(c) if romaji.starts_with(*c) => (
						input.kanas.get_current().sokuon_literal(),
						{
							input.romaji = utf8::slice_from(romaji, size);
							input
						},
						Self::Init,
					),
					_ => ("", input, Self::Fallback),
				},
				Self::Fallback => {
					let output = utf8::slice_to(romaji, size);
					input.romaji = utf8::slice_from(romaji, size);

					(output, input, Self::Init)
				}
				Self::Init => unreachable!("this has already been checked above"),
			}),
		}
	}

	fn get_size(&self) -> usize {
		match self {
			Self::SizeRouter => 0,
			Self::Init
			| Self::RawToggle
			| Self::KanaToggle
			| Self::RawText(_)
			| Self::Tiny
			| Self::VirtualSokuon
			| Self::Fallback => 1,
			Self::Short | Self::Sokuon | Self::SmallVowel | Self::Chouonpu => 2,
			Self::Medium => 3,
			Self::Long => 4,
			Self::Punctuation(size, _) => *size,
		}
	}
}

File: src/Fsm.zig
const std = @import("std");
const heap = std.heap;
const mem = std.mem;
const unicode = std.unicode;
const ascii = std.ascii;

const Config = @import("./Config.zig");
const Input = @import("./Input.zig");
const hiragana = @import("./hiragana.zig");
const katakana = @import("./katakana.zig");
const sokuon = @import("./sokuon.zig");
const chouonpu = @import("./chouonpu.zig");
const punctuation = @import("./punctuation.zig");
const helpers = @import("./helpers.zig");

const kana = enum {
    hiragana,
    katakana,

    fn get_syllabogram(self: kana, allocator: mem.Allocator, query: []const u8, cfg: Config) ?[]const u8 {
        const lowercase_query = ascii.allocLowerString(allocator, query) catch return null;
        defer allocator.free(lowercase_query);

        return switch (self) {
            .hiragana => hiragana.Syllabary.get(lowercase_query),
            .katakana => if (cfg.extended_katakana)
                katakana.ExtendedSyllabary.get(lowercase_query) orelse katakana.Syllabary.get(lowercase_query)
            else
                katakana.Syllabary.get(lowercase_query),
        };
    }

    fn get_sokuon(self: kana, allocator: mem.Allocator, query: []const u8) ?[]const u8 {
        const lowercase_query = ascii.allocLowerString(allocator, query) catch return null;
        defer allocator.free(lowercase_query);

        if (sokuon.Matches.has(lowercase_query)) {
            return switch (self) {
                .hiragana => hiragana.sokuon,
                .katakana => katakana.sokuon,
            };
        }

        return null;
    }

    fn get_small_vowels(self: kana, allocator: mem.Allocator, query: []const u8) ?[]const u8 {
        const lowercase_query = ascii.allocLowerString(allocator, query) catch return null;
        defer allocator.free(lowercase_query);

        return switch (self) {
            .hiragana => hiragana.SmallVowels.get(lowercase_query),
            .katakana => katakana.SmallVowels.get(lowercase_query),
        };
    }

    fn get_chouonpu(self: kana, allocator: mem.Allocator, query: []const u8, cfg: Config) ?[]const u8 {
        const lowercase_query = ascii.allocLowerString(allocator, query) catch return null;
        defer allocator.free(lowercase_query);

        return switch (self) {
            .hiragana => if (cfg.force_prolongation and chouonpu.Matches.has(lowercase_query)) chouonpu.Graph else null,
            .katakana => if (chouonpu.Matches.has(lowercase_query)) chouonpu.Graph else null,
        };
    }
};

const Fsm = @This();

pub const State = union(enum) {
    init,
    raw_toggle,
    kana_toggle,
    size_router,
    long,
    medium,
    short,
    tiny,
    sokuon,
    virt_sokuon,
    small_vowel,
    chouonpu,
    raw_text: u8,
    punctuation: struct { @"0": usize, @"1": *const State },
    fallback,
};

input: Input,
state: State = .init,
kanas: helpers.Switch(kana) = .{
    .data = [2]kana{
        .hiragana,
        .katakana,
    },
},
single_quotes: helpers.Switch([]const u8) = .{
    .data = punctuation.single_quotes,
},
double_quotes: helpers.Switch([]const u8) = .{
    .data = punctuation.double_quotes,
},
config: Config,
allocator: mem.Allocator,

pub fn init(allocator: mem.Allocator, text: []const u8, config: Config) Fsm {
    var fsm = Fsm{
        .allocator = allocator,
        .input = Input.init(text),
        .config = config,
    };

    if (config.start_with_katakana) {
        fsm.kanas.toggle();
    }

    return fsm;
}

/// Returns the computed string to be written and updates the state
/// machine with new input.
pub fn next(fsm: *Fsm) ?[]const u8 {
    switch (fsm.state) {
        .init => {
            if (fsm.input.len == 0) {
                return null;
            }

            fsm.state = .raw_toggle;
            return "";
        },
        .raw_toggle => {
            const size = 1;

            if (fsm.config.special_chars.raw_text) |toggle| {
                if (fsm.input.startsWith(toggle)) {
                    fsm.input.shrink(size, null);
                    fsm.state = .{ .raw_text = toggle };

                    return "";
                }
            }

            fsm.state = .kana_toggle;
            return "";
        },
        .raw_text => |toggle| {
            const size = 1;

            if (fsm.input.startsWith(toggle)) {
                fsm.input.shrink(size, null);
                fsm.state = .raw_toggle;

                return "";
            }

            if (fsm.input.view(size)) |char| {
                fsm.input.shrink(char.len, size);

                return char;
            }
        },
        .kana_toggle => {
            const size = 1;

            if (fsm.config.special_chars.kana) |toggle| {
                if (fsm.input.startsWith(toggle)) {
                    fsm.kanas.toggle();
                    fsm.input.shrink(size, null);
                    fsm.state = .init;

                    return "";
                }
            }

            fsm.state = .size_router;

            return "";
        },
        .size_router => {
            const n = fsm.input.len;

            fsm.state = if (n >= 4) .long else switch (n) {
                3 => .medium,
                2 => .sokuon,
                1 => .tiny,
                else => .init,
            };
            return "";
        },
        .long => return fsm.goto_next(4, .medium),
        .medium => return fsm.goto_next(3, .sokuon),
        .short => return fsm.goto_next(2, .tiny),
        .tiny => return fsm.goto_next(1, .virt_sokuon),
        .sokuon => {
            const size = 2;
            const query = fsm.input.view(size).?;
            const current_kana = fsm.kanas.get();

            if (current_kana.get_sokuon(fsm.allocator, query)) |result| {
                const str = fsm.input.view(size - 1).?;

                fsm.input.shrink(str.len, size - 1);
                fsm.state = .init;

                return result;
            }

            fsm.state = .small_vowel;

            return "";
        },
        .small_vowel => {
            if (fsm.config.special_chars.small_vowel) |trigger| {
                const current_kana = fsm.kanas.get();

                if (fsm.input.startsWith(trigger)) {
                    fsm.input.shrink(1, null);

                    const query = fsm.input.view(1).?;

                    if (current_kana.get_small_vowels(fsm.allocator, query)) |result| {
                        fsm.state = .chouonpu;

                        return result;
                    }
                }
            }

            fsm.state = .short;

            return "";
        },
        .punctuation => |values| {
            const size = values.@"0";
            const next_state = values.@"1";

            defer fsm.allocator.destroy(next_state);

            if (!fsm.config.parse_punctuation) {
                _ = fsm.input.view(size);
                fsm.state = next_state.*;

                return "";
            }

            if (fsm.input.startsWith('\'')) {
                fsm.input.shrink(1, null);
                fsm.state = .init;

                defer fsm.single_quotes.toggle();

                return fsm.single_quotes.get();
            }

            if (fsm.input.startsWith('"')) {
                fsm.input.shrink(1, null);
                fsm.state = .init;

                defer fsm.double_quotes.toggle();

                return fsm.double_quotes.get();
            }

            const query = fsm.input.view(size).?;
            if (punctuation.Marks.get(query)) |result| {
                fsm.input.shrink(size, null);
                fsm.state = .init;

                return result;
            }

            fsm.state = next_state.*;

            return "";
        },
        .chouonpu => {
            const size = 2;
            const query = fsm.input.view(size).?; // oo | o^
            const base = fsm.input.view(1).?; // o

            fsm.input.shrink(base.len, 1); // o | ^

            if (fsm.config.special_chars.reset_prolongation) |reset| {
                if (fsm.input.startsWith(reset)) {
                    fsm.input.shrink(1, null); // ?
                    fsm.state = .init;

                    return "";
                }
            }

            const current_kana = fsm.kanas.get();
            if (current_kana.get_chouonpu(fsm.allocator, query, fsm.config)) |result| {
                fsm.state = .chouonpu;

                return result;
            }

            fsm.state = .init;

            return "";
        },
        .virt_sokuon => {
            if (fsm.config.special_chars.virt_stop) |stop| {
                if (fsm.input.startsWith(stop)) {
                    fsm.input.shrink(1, null);
                    fsm.state = .init;

                    return switch (fsm.kanas.get()) {
                        .hiragana => hiragana.sokuon,
                        .katakana => katakana.sokuon,
                    };
                }
            }

            fsm.state = .fallback;

            return "";
        },
        .fallback => {
            const char = fsm.input.view(1).?;

            fsm.input.shrink(char.len, 1);
            fsm.state = .init;

            return char;
        },
    }

    return null;
}

fn goto_next(fsm: *Fsm, size: usize, next_state: State) ?[]const u8 {
    const query = fsm.input.view(size).?;
    const cur_kana = fsm.kanas.get();

    if (cur_kana.get_syllabogram(fsm.allocator, query, fsm.config)) |result| {
        const str = fsm.input.view(size - 1).?;

        fsm.input.shrink(str.len, size - 1);
        fsm.state = .chouonpu;

        return result;
    }

    const next_state_ptr = fsm.allocator.create(State) catch {
        return null;
    };
    errdefer fsm.allocator.destroy(next_state_ptr);

    next_state_ptr.* = next_state;
    fsm.state = .{ .punctuation = .{ .@"0" = size, .@"1" = next_state_ptr } };

    return "";
}

test "init" {}
